MaintenanceJobDesignDoc.md

```diff
@@ -0,0 +1,140 @@
+Maintenance Job Design Document
+===============================
+
+The (shared) object cache is growing to be large. This includes:
+
+* Number of loose objects.
+* Number of pack-files.
+* Size of pack-files.
+
+There are some things we can do to improve this situation, with varying levels
+of difficulty and benefit. Here are the four items I believe will have a strong
+impact on our data size and performance:
+
+I. Delete loose objects that exist in pack-files.
+II. Place loose objects into pack-files.
+III. Repack some prefetch packs into larger packs.
+IV. Delete packs with no objects referenced by the multi-pack-index.
+
+These are numbered by order of implementation, but they work together in an
+interesting way. If we place loose objects into pack-files (I), then the next
+time we delete loose objects (II) we will delete those objects that we just put
+into a pack-file. If we repack some prefetch packs into larger packs (III), the
+multi-pack-index will use the "newer" pack for all objects in those packs, and
+we can delete them (IV).
+
+In order to perform this maintenance without disrupting users, we can actually
+do these interactions in two independent jobs, with their sub-actions in a
+particular order:
+
+A. Loose Object Maintenance
+  1. Delete loose objects that exist in pack-files.
+  2. Place loose objects into pack-files.
+B. Prefetch Pack Maintenance
+  1. Delete packs with no objects referenced by the multi-pack-index.
+  2. Repack some prefetch packs into larger packs.
+
+Care must be taken to ensure that Git processes that initialized their object
+store before the maintenance   still succeed. Thus, we should never delete
+anything that those processes could be reading from. This is why we delete loose
+objects only if they are referenced by pack-files. This is why we delete
+prefetch packs only if they have no objects referenced by the multi-pack-index.
+Since a second run of these jobs could delete objects that were repacked in the
+previous run, we need to space our job runs such that no Git processes span both
+runs.

```

Comment:
Maybe we should have some tracking of the git processes that are running before we start and mark things for deletion and have something that only checks what can be deleted and makes sure the git processes are gone so we can know for sure that something can be deleted?

Reply:
Such a thing could be possible. I'll look into it. Seems like overkill for a job that will run at most once a day, but we can see how disruptive it would be to add a check after the job runs.